[hipdnn][ALMIOPEN-868] Added FP8 data types support

a-sidorova opened the pull request:
## Motivation

Updated `hipdnn` to have fp8 operation support within the ticket ALMIOPEN-868.

## Technical Details

The PR implements support of FP8 data type (`fp8_e4m3`) to hipDNN frontend, data sdk, test sdk using `__hip_fp8_e4_m3` classes from HIP. Implemented the files `UtilsFp8.hpp` for work with these data types in hipDNN. The changes are based on https://github.com/ROCm/rocm-libraries/pull/3510.

Also the PR adds missed overrided comparison function for `half` data type. Without these changes, if there is comparison of half values, now the build failed because the compiler now finds ambiguous overloads (there are multiple options: to be converted to float or new fp8).

Depends on the fix: https://github.com/ROCm/rocm-systems/pull/2448

## Test Plan

Added utility FP8 tests, convolution and batchnorm tests with FP8. As far as I understand, we cannot add Pointwise tests with such data types due to unimplemented operators +,-,etc for this data type in HIP.

## Test Result

<!-- Briefly summarize test outcomes. -->

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

a-sidorova commented on projects/hipdnn/sdk/include/hipdnn_sdk/utilities/UtilsFp8.hpp:
```diff
@@ -0,0 +1,170 @@
+    return {static_cast<float>(val)};
+}
+
+// NOLINTBEGIN(modernize-avoid-c-arrays): False positive on template SFINAE expressions
```
False positive:
```
/opt/rocm-libraries/projects/hipdnn/sdk/include/hipdnn_sdk/utilities/UtilsFp8.hpp:34:63: error: do not declare C-style arrays, use 'std::array' instead [modernize-avoid-c-arrays,-warnings-as-errors]
   34 |                                && std::is_same_v<std::decay_t<U>, hip_fp8_e4m3>,
      |                                                               ^
53217 warnings generated.
```

a-sidorova commented:
At the moment TheRock CI jobs are failed with the following errors 🤔 :

```
2025-12-23T11:42:54.7076881Z [hipDNN] /__w/rocm-libraries/rocm-libraries/TheRock/build/core/clr/dist/include/hip/amd_detail/amd_hip_ocp_types.h:87:2: error: "Only supported by HIPCC or GCC >= 13."
2025-12-23T11:42:54.7078227Z [hipDNN]    87 | #error "Only supported by HIPCC or GCC >= 13."
2025-12-23T11:42:54.7078784Z [hipDNN]       |  ^
2025-12-23T11:42:54.7079961Z [hipDNN] /__w/rocm-libraries/rocm-libraries/TheRock/build/core/clr/dist/include/hip/amd_detail/amd_hip_ocp_types.h:90:22: error: use of undeclared identifier '__amd_uintx2_storage_t'
2025-12-23T11:42:54.7081536Z [hipDNN]    90 | static_assert(sizeof(__amd_uintx2_storage_t) == sizeof(__amd_fp8x8_storage_t));
2025-12-23T11:42:54.7082301Z [hipDNN]       |                      ^~~~~~~~~~~~~~~~~~~~~~
2025-12-23T11:42:54.7083640Z [hipDNN] /__w/rocm-libraries/rocm-libraries/TheRock/build/core/clr/dist/include/hip/amd_detail/amd_hip_ocp_types.h:90:56: error: use of undeclared identifier '__amd_fp8x8_storage_t'
2025-12-23T11:42:54.7085220Z [hipDNN]    90 | static_assert(sizeof(__amd_uintx2_storage_t) == sizeof(__amd_fp8x8_storage_t));
2025-12-23T11:42:54.7086012Z [hipDNN]       |                                                        ^~~~~~~~~~~~~~~~~~~~~
2025-12-23T11:42:54.7087111Z [hipDNN] In file included from /__w/rocm-libraries/rocm-libraries/projects/hipdnn/frontend/tests/TestBatchnormBackwardNode.cpp:4:
2025-12-23T11:42:54.7088675Z [hipDNN] In file included from /__w/rocm-libraries/rocm-libraries/projects/hipdnn/frontend/include/hipdnn_frontend/Error.hpp:6:
2025-12-23T11:42:54.7090204Z [hipDNN] In file included from /__w/rocm-libraries/rocm-libraries/projects/hipdnn/frontend/include/hipdnn_frontend/Types.hpp:11:
2025-12-23T11:42:54.7091732Z [hipDNN] In file included from /__w/rocm-libraries/rocm-libraries/projects/hipdnn/sdk/include/hipdnn_sdk/utilities/UtilsBfp8.hpp:6:
2025-12-23T11:42:54.7093264Z [hipDNN] In file included from /__w/rocm-libraries/rocm-libraries/TheRock/build/core/clr/dist/include/hip/hip_fp8.h:30:
2025-12-23T11:42:54.7094712Z [hipDNN] In file included from /__w/rocm-libraries/rocm-libraries/TheRock/build/core/clr/dist/include/hip/amd_detail/amd_hip_fp8.h:68:
2025-12-23T11:42:54.7096495Z [hipDNN] /__w/rocm-libraries/rocm-libraries/TheRock/build/core/clr/dist/include/hip/amd_detail/amd_hip_mx_common.h:49:14: error: no member named 'data' in '__half_raw'
2025-12-23T11:42:54.7098105Z [hipDNN]    49 |   return tmp.data;
2025-12-23T11:42:54.7098574Z [hipDNN]       |          ~~~ ^
2025-12-23T11:42:54.7099978Z [hipDNN] /__w/rocm-libraries/rocm-libraries/TheRock/build/core/clr/dist/include/hip/amd_detail/amd_hip_mx_common.h:52:35: error: unknown type name '__amd_fp16x2_storage_t'
2025-12-23T11:42:54.7101510Z [hipDNN]    52 | __host__ __device__ static inline __amd_fp16x2_storage_t half2_to_f16x2(const __half2 val) {
2025-12-23T11:42:54.7102315Z [hipDNN]       |                                   ^
2025-12-23T11:42:54.7103542Z [hipDNN] /__w/rocm-libraries/rocm-libraries/TheRock/build/core/clr/dist/include/hip/amd_detail/amd_hip_mx_common.h:54:14: error: no member named 'data' in '__half2_raw'
2025-12-23T11:42:54.7104765Z [hipDNN]    54 |   return tmp.data;
2025-12-23T11:42:54.7105206Z [hipDNN]       |          ~~~ ^
2025-12-23T11:42:54.7108045Z [hipDNN] /__w/rocm-libraries/rocm-libraries/TheRock/build/core/clr/dist/include/hip/amd_detail/amd_hip_mx_common.h:66:35: error: unknown type name '__amd_bf16x2_storage_t'; did you mean '__amd_bf16_storage_t'?
2025-12-23T11:42:54.7109861Z [hipDNN]    66 | __host__ __device__ static inline __amd_bf16x2_storage_t hipbf162_to_bf16x2(const __hip_bfloat162 val) {
2025-12-23T11:42:54.7110704Z [hipDNN]       |                                   ^
2025-12-23T11:42:54.7111926Z [hipDNN] /__w/rocm-libraries/rocm-libraries/TheRock/build/core/clr/dist/include/hip/amd_detail/amd_hip_ocp_types.h:51:16: note: '__amd_bf16_storage_t' declared here
2025-12-23T11:42:54.7113596Z [hipDNN]    51 | typedef __bf16 __amd_bf16_storage_t;
2025-12-23T11:42:54.7114096Z [hipDNN]       |                ^
2025-12-23T11:42:54.7115069Z [hipDNN] In file included from /__w/rocm-libraries/rocm-libraries/projects/hipdnn/frontend/tests/TestBatchnormBackwardNode.cpp:4:
2025-12-23T11:42:54.7116576Z [hipDNN] In file included from /__w/rocm-libraries/rocm-libraries/projects/hipdnn/frontend/include/hipdnn_frontend/Error.hpp:6:
2025-12-23T11:42:54.7118076Z [hipDNN] In file included from /__w/rocm-libraries/rocm-libraries/projects/hipdnn/frontend/include/hipdnn_frontend/Types.hpp:11:
2025-12-23T11:42:54.7119578Z [hipDNN] In file included from /__w/rocm-libraries/rocm-libraries/projects/hipdnn/sdk/include/hipdnn_sdk/utilities/UtilsBfp8.hpp:6:
2025-12-23T11:42:54.7121013Z [hipDNN] In file included from /__w/rocm-libraries/rocm-libraries/TheRock/build/core/clr/dist/include/hip/hip_fp8.h:30:
2025-12-23T11:42:54.7122503Z [hipDNN] In file included from /__w/rocm-libraries/rocm-libraries/TheRock/build/core/clr/dist/include/hip/amd_detail/amd_hip_fp8.h:68:
2025-12-23T11:42:54.7124499Z [hipDNN] /__w/rocm-libraries/rocm-libraries/TheRock/build/core/clr/dist/include/hip/amd_detail/amd_hip_mx_common.h:67:51: error: unknown type name '__amd_bf16x2_storage_t'; did you mean '__amd_bf16_storage_t'?
2025-12-23T11:42:54.7126165Z [hipDNN]    67 |   static_assert(sizeof(__hip_bfloat162) == sizeof(__amd_bf16x2_storage_t));
2025-12-23T11:42:54.7126904Z [hipDNN]       |                                                   ^
2025-12-23T11:42:54.7128166Z [hipDNN] /__w/rocm-libraries/rocm-libraries/TheRock/build/core/clr/dist/include/hip/amd_detail/amd_hip_ocp_types.h:51:16: note: '__amd_bf16_storage_t' declared here
2025-12-23T11:42:54.7129420Z [hipDNN]    51 | typedef __bf16 __amd_bf16_storage_t;
2025-12-23T11:42:54.7129925Z [hipDNN]       |                ^
2025-12-23T11:42:54.7130899Z [hipDNN] In file included from /__w/rocm-libraries/rocm-libraries/projects/hipdnn/frontend/tests/TestBatchnormBackwardNode.cpp:4:
2025-12-23T11:42:54.7132435Z [hipDNN] In file included from /__w/rocm-libraries/rocm-libraries/projects/hipdnn/frontend/include/hipdnn_frontend/Error.hpp:6:
2025-12-23T11:42:54.7133929Z [hipDNN] In file included from /__w/rocm-libraries/rocm-libraries/projects/hipdnn/frontend/include/hipdnn_frontend/Types.hpp:11:
2025-12-23T11:42:54.7135458Z [hipDNN] In file included from /__w/rocm-libraries/rocm-libraries/projects/hipdnn/sdk/include/hipdnn_sdk/utilities/UtilsBfp8.hpp:6:
2025-12-23T11:42:54.7137300Z [hipDNN] In file included from /__w/rocm-libraries/rocm-libraries/TheRock/build/core/clr/dist/include/hip/hip_fp8.h:30:
2025-12-23T11:42:54.7138800Z [hipDNN] In file included from /__w/rocm-libraries/rocm-libraries/TheRock/build/core/clr/dist/include/hip/amd_detail/amd_hip_fp8.h:68:
2025-12-23T11:42:54.7141087Z [hipDNN] /__w/rocm-libraries/rocm-libraries/TheRock/build/core/clr/dist/include/hip/amd_detail/amd_hip_mx_common.h:67:17: error: static assertion failed due to requirement 'sizeof(__hip_bfloat162) == sizeof(__bf16)'
2025-12-23T11:42:54.7142800Z [hipDNN]    67 |   static_assert(sizeof(__hip_bfloat162) == sizeof(__amd_bf16x2_storage_t));
2025-12-23T11:42:54.7143580Z [hipDNN]       |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2025-12-23T11:42:54.7144868Z [hipDNN] /__w/rocm-libraries/rocm-libraries/TheRock/build/core/clr/dist/include/hip/amd_detail/amd_hip_mx_common.h:67:41: note: expression evaluates to '4 == 2'
2025-12-23T11:42:54.7146286Z [hipDNN]    67 |   static_assert(sizeof(__hip_bfloat162) == sizeof(__amd_bf16x2_storage_t));
2025-12-23T11:42:54.7147073Z [hipDNN]       |                 ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2025-12-23T11:42:54.7148564Z [hipDNN] /__w/rocm-libraries/rocm-libraries/TheRock/build/core/clr/dist/include/hip/amd_detail/amd_hip_mx_common.h:70:5: error: unknown type name '__amd_bf16x2_storage_t'; did you mean '__amd_bf16_storage_t'?
2025-12-23T11:42:54.7150334Z [hipDNN]    70 |     __amd_bf16x2_storage_t bf16;
2025-12-23T11:42:54.7150812Z [hipDNN]       |     ^
2025-12-23T11:42:54.7151897Z [hipDNN] /__w/rocm-libraries/rocm-libraries/TheRock/build/core/clr/dist/include/hip/amd_detail/amd_hip_ocp_types.h:51:16: note: '__amd_bf16_storage_t' declared here
2025-12-23T11:42:54.7153153Z [hipDNN]    51 | typedef __bf16 __amd_bf16_storage_t;
```

The include `<hip/hip_fp8.h>` leads to these compilation errors.

adickin-amd commented:
Thank you for adding the support for these types!

Looks like this PR has all the int8 changes as well as the bfp8 and fp8 changes.  Can we split up this PR so that its bringing in each type 1 at a time.  Its a bit more process but will speed up our review cycles.  Since there is a build break, itll also be easier to debug with less changed code.

- Does the bfp8 header have the same build breaks as the fp8 header does?  Wondering if we can bring in the bfp8 work then spend the time fixing up fp8.  We did have lots of issues when bringing in fp16/bfp16 and did have to add some extra cmake defines to pull things in nicely without needing hip.  There is also some changes being made to the bfp16/fp16 headers to make it easier in the future.(https://github.com/ROCm/rocm-systems/pull/882)

a-sidorova commented:
> Thank you for adding the support for these types!
> 
> Looks like this PR has all the int8 changes as well as the bfp8 and fp8 changes. Can we split up this PR so that its bringing in each type 1 at a time. Its a bit more process but will speed up our review cycles. Since there is a build break, itll also be easier to debug with less changed code.
> 
> * Does the bfp8 header have the same build breaks as the fp8 header does?  Wondering if we can bring in the bfp8 work then spend the time fixing up fp8.  We did have lots of issues when bringing in fp16/bfp16 and did have to add some extra cmake defines to pull things in nicely without needing hip.  There is also some changes being made to the bfp16/fp16 headers to make it easier in the future.([SWDEV-553774 - enable fp16 header for clang rocm-systems#882](https://github.com/ROCm/rocm-systems/pull/882))

Thank you for the recommendation! I have split the implementations in two PRs: https://github.com/ROCm/rocm-libraries/pull/3525 and https://github.com/ROCm/rocm-libraries/pull/3545.

As for compilation problems in CI, this errors occurs because of include the file `hip/hip_fp8.h`. This file is needed for fp8 and bfp8. I tried to fix it and create the PR to rocm-systems - https://github.com/ROCm/rocm-systems/pull/2448

a-sidorova commented on projects/hipdnn/data_sdk/include/hipdnn_data_sdk/utilities/UtilsFp16.hpp:
```diff
@@ -14,6 +14,43 @@ inline __HOST_DEVICE__ half operator""_h(long double value)
     return {static_cast<float>(value)};
 }
 
+inline __HOST_DEVICE__ half operator-(half a)
+{
+    auto r = static_cast<__half_raw>(a);
+    r.x ^= 0x8000u;
+    return r;
+}
+
+inline __HOST_DEVICE__ bool operator==(half a, half b)
+{
+    return static_cast<__half_raw>(a).x == static_cast<__half_raw>(b).x;
+}
+
+inline __HOST_DEVICE__ bool operator!=(half a, half b)
+{
+    return static_cast<__half_raw>(a).x != static_cast<__half_raw>(b).x;
+}
+
+inline __HOST_DEVICE__ bool operator<(half a, half b)
+{
+    return static_cast<__half_raw>(a).x < static_cast<__half_raw>(b).x;
+}
+
+inline __HOST_DEVICE__ bool operator>(half a, half b)
+{
+    return static_cast<__half_raw>(a).x > static_cast<__half_raw>(b).x;
+}
+
+inline __HOST_DEVICE__ bool operator<=(half a, half b)
+{
+    return static_cast<__half_raw>(a).x <= static_cast<__half_raw>(b).x;
+}
+
+inline __HOST_DEVICE__ bool operator>=(half a, half b)
+{
+    return static_cast<__half_raw>(a).x >= static_cast<__half_raw>(b).x;
+}
+
```
Added missed overrided operators for `__half` from [ `hip_fp16_gcc.h`](https://github.com/a-sidorova/rocm-systems/blob/f7338717aeb56fbe394c0683539a0cf2f7f3ded7/projects/clr/hipamd/include/hip/amd_detail/hip_fp16_gcc.h)

Without these changes, if there is comparison of `half` values, now the build failed because the compiler now finds ambiguous overloads (there are multiple options: to be converted to float or fp8):
```
/opt/rocm-libraries/projects/hipdnn/test_sdk/include/hipdnn_test_sdk/utilities/cpu_graph_executor/BatchnormTrainSignatureKey.hpp:108:9: note: in instantiation of function template specialization 'hipdnn_test_sdk::utilities::BatchnormTrainSignatureKey::addPlanBuilder<hipdnn_data_sdk::data_objects::DataType::HALF, hipdnn_data_sdk::data_objects::DataType::HALF, hipdnn_data_sdk::data_objects::DataType::HALF, hipdnn_data_sdk::data_objects::DataType::HALF, hipdnn_data_sdk::data_objects::DataType::HALF>' requested here
  108 |         addPlanBuilder<hipdnn_data_sdk::data_objects::DataType::HALF,
      |         ^
/opt/rocm-libraries/projects/hipdnn/data_sdk/include/hipdnn_data_sdk/utilities/UtilsFp8.hpp:27:29: note: candidate function
   27 | inline __HOST_DEVICE__ bool operator==(const hip_fp8_e4m3& a, const hip_fp8_e4m3& b)
      |                             ^
/opt/rocm-libraries/projects/hipdnn/test_sdk/include/hipdnn_test_sdk/utilities/CpuFpReferenceBatchnorm.hpp:177:28: note: built-in candidate operator==(float, float)
  177 |                     = (nhw == one) ? channelVariance
```

SamuelReeder commented on projects/hipdnn/data_sdk/include/hipdnn_data_sdk/utilities/FlatbufferUtils.hpp:
```diff
@@ -76,6 +76,18 @@ TargetType extractValueFromTensorValue(const data_objects::TensorAttributesT& te
             return static_cast<TargetType>(val->value());
         }
         break;
+    case data_objects::DataType::INT8:
+        if(auto val = tensorAttr.value.AsFloat8Value())
+        {
+            return static_cast<TargetType>(val->value());
+        }
+        break;
```
Looks like some changes from the int8 PR leaked in. Not a big deal for this, but it would reduce the diff and be easier to review if you branched off `develop` for any new PR.

a-sidorova replied:
I specially make new branches based on the previous (like a `matreshka` :)) to make the rebasing on the master in the future easier. Since we update enum classes here with explicit values. 

Reviewers can take just one last commit for review - I hope that then the review should not be difficult.
Anyway I hope that we will int8 support soon and the review process will be easier :)

SamuelReeder commented on projects/hipdnn/data_sdk/include/hipdnn_data_sdk/utilities/UtilsFp8.hpp:
```diff
@@ -0,0 +1,126 @@
+    return val;
+}
+
+inline __HOST_DEVICE__ bool operator==(hip_fp8_e4m3 a, hip_fp8_e4m3 b)
+{
+    return a.__x == b.__x;
+}
+
+inline __HOST_DEVICE__ bool operator!=(hip_fp8_e4m3 a, hip_fp8_e4m3 b)
+{
+    return a.__x != b.__x;
+}
+
+inline __HOST_DEVICE__ bool operator>(hip_fp8_e4m3 a, hip_fp8_e4m3 b)
+{
+    return a.__x > b.__x;
+}
+
+inline __HOST_DEVICE__ bool operator>=(hip_fp8_e4m3 a, hip_fp8_e4m3 b)
+{
+    return a.__x >= b.__x;
+}
+
+inline __HOST_DEVICE__ bool operator<(hip_fp8_e4m3 a, hip_fp8_e4m3 b)
+{
+    return a.__x < b.__x;
+}
+
+inline __HOST_DEVICE__ bool operator<=(hip_fp8_e4m3 a, hip_fp8_e4m3 b)
+{
+    return a.__x <= b.__x;
+}
```
Should the params in all these comparisons be casted to float first? `__x` is of type `__hip_fp8_storage_t` ([source](https://github.com/ROCm/rocm-systems/blob/develop/projects/clr/hipamd/include/hip/amd_detail/amd_hip_fp8.h#L1077)), which is an unsigned char ([source](https://github.com/ROCm/rocm-systems/blob/develop/projects/clr/hipamd/include/hip/amd_detail/amd_hip_fp8.h#L139-L143)). Therefore, it seems this would use unsigned integer comparison where a negative number with a leading sign bit of 1 would be evaluated as greater than any positive number.

a-sidorova replied:
You are totally right!! Thanks for the catch! It was caught by added new tests also. The same mistake was for fp16.

Added casting to float everywhere.

SamuelReeder commented on projects/hipdnn/data_sdk/include/hipdnn_data_sdk/utilities/UtilsFp8.hpp:
```diff
@@ -0,0 +1,126 @@
+    return a.__x <= b.__x;
+}
+
+namespace hipdnn_sdk::utilities::fp8
```
This namespace should be `hipdnn_data_sdk::utilities::fp8`.

a-sidorova replied:
My bad, missed after rebasing. Thanks for the catch!

SamuelReeder commented on projects/hipdnn/data_sdk/include/hipdnn_data_sdk/utilities/UtilsFp8.hpp:
```diff
@@ -0,0 +1,126 @@
+namespace std
+{
+
+inline __HOST_DEVICE__ hip_fp8_e4m3 fabs(hip_fp8_e4m3 num)
+{
+    return hipdnn_sdk::utilities::fp8::fp8abs(num);
+}
+
+inline __HOST_DEVICE__ hip_fp8_e4m3 abs(hip_fp8_e4m3 num)
+{
+    return hipdnn_sdk::utilities::fp8::fp8abs(num);
+}
+
+inline __HOST_DEVICE__ hip_fp8_e4m3 max(hip_fp8_e4m3 a, hip_fp8_e4m3 b)
+{
+    return hipdnn_sdk::utilities::fp8::fp8max(a, b);
+}
```
These utilities are under the `hipdnn_data_sdk` namespace. Sorry for the rename :)

a-sidorova replied:
Done, thanks!

SamuelReeder commented on projects/hipdnn/data_sdk/tests/utilities/TestFloats.cpp:
```diff
@@ -45,3 +46,27 @@ TEST(TestUtilsBfp16, Max)
     EXPECT_EQ(std::max(a, b), 2.0_bf);
     EXPECT_EQ(std::max(b, a), 2.0_bf);
 }
+
+TEST(TestUtilsFp8, BasicUsage)
+{
+    hip_fp8_e4m3 fp8 = 1.0_fp8;
+    EXPECT_EQ(fp8, 1.0_fp8);
+}
+
+TEST(TestUtilsFp8, Fabs)
+{
+    EXPECT_EQ(std::fabs(-1.0_fp8), 1.0_fp8);
+    EXPECT_EQ(std::fabs(1.0_fp8), 1.0_fp8);
+}
+
+TEST(TestUtilsFp8, Max)
+{
+    hip_fp8_e4m3 a = 1.0_fp8;
+    hip_fp8_e4m3 b = 2.0_fp8;
+    hip_fp8_e4m3 nan = hipdnn_sdk::utilities::fp8::uchar_as_fp8(0x7F);
+    EXPECT_EQ(std::max(a, b), 2.0_fp8);
+    EXPECT_EQ(std::max(b, a), 2.0_fp8);
+    EXPECT_EQ(std::max(a, nan), 1.0_fp8);
+    EXPECT_EQ(std::max(nan, b), 2.0_fp8);
+    EXPECT_EQ(std::max(nan, nan), nan);
+}
```
I know we don't currently test the comparison operators for the others, but I think it would be good to start doing that. Maybe add a test case that verifies some basic comparisons.

a-sidorova replied:
Good idea! Added tests for fp16, bf16 and fp8. Thanks to these tests I found mistake in my comparison operators for fp8.

Thanks!

SamuelReeder commented on projects/hipdnn/test_sdk/include/hipdnn_test_sdk/utilities/pointwise/UnaryOperationFunctors.hpp:
```diff
@@ -92,7 +92,7 @@ struct Negation
     auto operator()(const X& x) const -> X
     {
-        return -x;
+        return static_cast<X>(-x);
```
Ah, might be implicit conversion. Nice catch!

a-sidorova commented:
> Generally looks good! Just left a few comments to address.

@SamuelReeder thank you so much for review! I have applied your comments. May I ask you to review one more time when you have time?

SamuelReeder commented on projects/hipdnn/data_sdk/include/hipdnn_data_sdk/utilities/UtilsFp16.hpp:
```diff
@@ -23,32 +23,34 @@ inline __HOST_DEVICE__ half operator-(half a)
 {
-    return static_cast<__half_raw>(a).x >= static_cast<__half_raw>(b).x;
+    return static_cast<float>(a) >= static_cast<float>(b);
+    ;
```
Looks like there's a couple stray semicolons here.

a-sidorova replied:
Oh, my bad. Smth went wrong during refactoring. Thanks for the catch!

a-sidorova commented:
> LGTM! Thank you for adding the additional test cases.
> 
> There's just a couple stray semicolons to remove before merging.

@SamuelReeder thanks a lot for the review! I have fixed the last comment in commit https://github.com/ROCm/rocm-libraries/pull/3525/commits/0876322f7b16fcd5a731f01fe85d9e47a7e94b08.

As for merge, we still have [build problems](https://github.com/ROCm/rocm-libraries/actions/runs/20589849027/job/59132971920?pr=3525) in the Rock CI which should be fixed by https://github.com/ROCm/rocm-systems/pull/2448. Looks like the reviewer @b-sumner  is OOO so probably the merge of fix will take some time :)

sshlyapn commented on projects/hipdnn/data_sdk/include/hipdnn_data_sdk/utilities/FlatbufferUtils.hpp:
```diff
@@ -82,6 +82,12 @@ TargetType extractValueFromTensorValue(const data_objects::TensorAttributesT& te
             return static_cast<TargetType>(val->value());
         }
         break;
+    case data_objects::DataType::FP8_E4M3:
+        if(auto val = tensorAttr.value.AsFloat8Value())
+        {
+            return static_cast<TargetType>(val->value());
+        }
+        break;
```
@SamuelReeder, can you please comment on this casting?
Shouldn't it be implemented like this to properly cast uint8_t raw value to arbitrary TargetType?
```
            auto fp8 = hipdnn_sdk::utilities::fp8::uchar_as_fp8(val->value());
            return static_cast<TargetType>(static_cast<float>(fp8));
```

SamuelReeder replied:
Yes `uint8_t` types should be reinterpreted. Good catch!

a-sidorova replied:
@sshlyapn thanks for the catch!

guys, fixed in https://github.com/ROCm/rocm-libraries/pull/3525/commits/9e2f13f9602d00da46afe0dcb90ec6e8ecf6cdbf
